### PR TITLE
Startup: one config tree, every case built once, the startup clock

### PR DESCRIPTION
**Pull request by vboussot**
## Summary

A workflow's startup is spent resolving its config and building its cases before any patch is read; this branch removes the repeated work on that path and reports where the rest goes. `strict_config` loads the config tree once per workflow build, every `Config` context inside it reads that tree and folds what it set back on exit, and the block writes the file once when it ends, in the order the per-context write-back produced (a context outside a block keeps loading and writing the file itself). A float validation split builds every case once, in run order with the training draws, and cuts the training partition as that build's head and the validation partition as its tail (validation without the draws is rebuilt without them, its share only). A root's `Dataset` is kept across the resolves of one run (the sizing pass, the selection, an OOM re-plan), so each header is parsed once. `rank_pool()` records the share it was built for and rebuilds at another share, which the sequential Python API reaches. `--version` resolves the version only when asked. The cohort walk seeds its intersections from the first group and the first root whatever they hold, refuses a naming subset no group can serve with the subset refusal instead of the group-intersection error or a `KeyError`, mirrors the file-before-slice test between `Subset._included_names` and `_resolve_selector`, and computes `required_names()` once per entry point. Rank 0 prints one `[KonfAI] startup` line in the sweep's format, `build + cases + grids + model + checkpoint + setup + launch + other`, with disjoint phases, the cross-process stamps in `time.time`, and the clock restarted around each Python API call.

## Measured

Config: scratch harness counting `_load_tree` and `yaml.dump` calls, 3 runs each (build = the workflow builder on a scratch copy).

| Config | Loads + dumps | Build |
|---|---|---|
| Synthesis/Config_GAN.yml | 153 + 76 (2.96 s) -> 1 + 1 (0.03 s) | 4.60-4.97 s -> 1.66-2.64 s (median 4.70 -> 2.04) |
| Synthesis/Config.yml | 103 + 51 (1.24 s) -> 1 + 1 | median 3.19 -> 2.27 s |
| Segmentation/Config.yml | 67 + 33 (0.61 s) -> 1 + 1 | median 1.49 -> 1.05 s |
| Segmentation/Prediction.yml | 33 + 16 (0.14 s) -> 1 + 1 | median 0.20 -> 0.05 s |

Float split: scratch harness on a synthetic cohort (2 groups, Flip nb=1, validation 0.2, torch seed 0, 3 runs).
- 10 cases: managers built 40 -> 20 (unaugmented validation: 40 -> 24)
- 200 cases: 800 -> 400, `prepare()` 0.349-0.359 s -> 0.296-0.337 s (the cohort listing dominates there; the count is the evidence)

Kept `Dataset`: scratch harness over `DataMetric.prepare()` on a synthetic mha cohort, two groups, 5 and 3 runs.
- 10 cases: Dataset objects 2 -> 1, headers parsed 40 -> 20 (0 -> 20 hits), prepare 0.034-0.036 s -> 0.019-0.020 s
- 200 cases: 2 -> 1, parsed 800 -> 400 (0 -> 400 hits), prepare 0.722-0.816 s -> 0.363-0.412 s

`--version`: `importlib.metadata.version("konfai")` measured 16.8 ms cold in the dev env (1.2 ms once warm), paid only when `--version` is given.

Startup line: Segmentation example on a synthetic cohort, inline on CPU, 3 runs each, `utils/config.py` at 6af1202 against HEAD.
- before: startup 1.3 / 1.4 / 1.5 s = build 1.0-1.1 + model 0.3 + launch 0.1
- after: startup 0.6 / 0.8 / 0.6 s = build 0.5-0.7 + model 0.0 + launch 0.1

**Values:** the perf changes move nothing. The resolved config is byte-identical on the 12 example configs (sha256 before and after; Config_GAN.yml and Transform.yml gain keys, which is where the order shows); the float split's draws are sha-identical, with the same names and indices in both partitions; the kept Dataset changes no value. The cohort-walk fixes change which cases a run selects where the groups' or roots' order decided it, what a names file spelled like a slice selects, and which error a naming subset no group can serve fails on.

**Tests:** `tests/unit/test_config.py` pins the one-tree read and write against the per-context path kept as the oracle (including a shipped config resolving to the same bytes, a refusing block leaving what it bound, and a file a context refused not being created); `test_data_manager.py` pins the single build with the partitions cut from it and the roots kept across an evaluation's two resolves; `test_cohort_walk.py` the first-group and first-root seeding, the subset refusal whatever the order, and a names file spelled like a slice; `test_runtime.py` the pool rebuild, the startup line's disjoint phases and rank 0's report; `test_main_cli.py` the version looked up only when asked; the `_resolve_dataset_sources` stub in `test_memory_budget.py` takes the required names as an argument.

**Stack:** Stacked on `pr/1.8.2-02-storage`; merge after it. Before deleting this branch, retarget the next PR of the stack.

## Summary by CodeRabbit

- **New Features**
  - Added startup timing visibility for workflow construction, setup, model loading, and launch phases.
  - The CLI now reports the installed version when `--version` is requested.
  - Dataset selection now supports subset-based requests and clearer handling of missing cases.

- **Bug Fixes**
  - Improved dataset reuse and partitioning consistency across repeated operations.
  - Configuration updates now use safer atomic writes and consolidated nested changes.
  - Clarified errors for invalid or empty dataset selections.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

File: konfai/trainer.py [hunk 1]
```diff
@@ -876,9 +878,10 @@ def setup(self, world_size: int):
                 statistics_path.unlink()
 
         state_dict = {}
-        if state != State.TRAIN:
-            state_dict = self._load()
-        self.model.load(state_dict, init=True, ema=False, override_lr=self.override_lr)
+        with startup_clock().phase("checkpoint"):
+            if state != State.TRAIN:
+                state_dict = self._load()
+            self.model.load(state_dict, init=True, ema=False, override_lr=self.override_lr)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/trainer.py:884]: Fixed. The phase closed before `AveragedModel` and `model_ema.module.load(...)`, so on a RESUME with `ema_decay` that time was reported under `setup`. Pinned by test_restoring_the_ema_weights_is_charged_to_the_checkpoint_phase.

File: konfai/utils/config.py [hunk 2]
```diff
@@ -154,11 +231,19 @@ def strict_config(root: str, refuse: bool = True) -> Iterator[None]:
...
+        _shared_trees.append(shared)
     try:
         yield
     finally:
         _ledgers.remove(ledger)
+        if shared is not None:
+            _shared_trees.remove(shared)
+            # Written whatever ended the block: what the contexts bound is on disk, as when each
+            # context wrote its own level.
+            shared.flush()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/config.py:246]: Fixed, and it was losing writes: each block built its own tree from disk, so the outer block's flush (of a tree read before the inner block existed) landed last and took the inner block's resolved root back to what the file held. A nested block now binds to the tree already held: one read, one write by the outermost. Pinned by test_a_block_opened_inside_another_one_on_the_same_file_keeps_both_blocks_writes.

File: konfai/utils/runtime.py [hunk 3]
```diff
@@ -1039,13 +1119,19 @@ def rank_pool() -> ThreadPoolExecutor | None:
 
     ``None`` at a share of one, where the work stays on the calling thread.
     """
-    global _rank_pool
+    global _rank_pool, _rank_pool_share
     workers = rank_cpu_share()
     if workers <= 1:
         return None
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/runtime.py:1125]: Fixed. The early return sat before the lock, so a pool built for a wider share kept its idle threads for the rest of the process, which the Python API reaches in one process (multi-rank build, then an inline single-rank workflow). Pinned by an extension of test_the_rank_pool_is_rebuilt_when_the_share_changes.

File: tests/unit/test_config.py [hunk 4]
```diff
@@ -725,3 +726,241 @@ def test_strict_config_reports_a_yaml_syntax_error_as_a_config_error(write_confi
...
+    "Transformer": ("konfai.transformer", "Transformer", "TRANSFORM", {"KONFAI_TRANSFORMS_DIRECTORY"}),
+}
+
+
+def _shipped_workflow_configs() -> list[str]:
+    return sorted(
+        str(path.relative_to(_EXAMPLES))
+        for path in _EXAMPLES.glob("*/*.yml")
+        if next(iter(ruamel.yaml.YAML().load(path.read_text(encoding="utf-8")))) in _WORKFLOWS
+    )
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_config.py:909]: Done. The scan feeds a `parametrize`, so an empty or sequence-topped example would have failed collection of the whole module; it now keeps files whose top level is a mapping opening on a workflow root, which is the same twelve configs.

File: tests/unit/test_data_manager.py [hunk 5]
```diff
@@ -285,6 +283,63 @@ def test_data_train_prepare_skips_validation_augmentation_layout_when_disabled(t
...
+    constructed: list[tuple[str, int]] = []
+
+    class CountingManager(DatasetManager):
+        def __init__(self, index: int, group_src: str, group_dest: str, name: str, *args, **kwargs) -> None:
+            constructed.append((name, index))
+            super().__init__(index, group_src, group_dest, name, *args, **kwargs)
+
+    monkeypatch.setattr(data_manager_module, "DatasetManager", CountingManager)
+    monkeypatch.delenv("KONFAI_config_file")  # the draw binds its own defaults
+    monkeypatch.setenv("KONFAI_ROOT", "Trainer")
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_data_manager.py:312]: Not needed: `tests/conftest.py::_konfai_env` is autouse and already sets both with `monkeypatch.setenv` for every test. That is also why `monkeypatch.delenv("KONFAI_config_file")` works here without `raising=False`. Repeating it per test would be a second spelling of one mechanism.
